### DISCUSSION TITLE
A Hint on discriminated unions?

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#370] [Q&A] [answered]
I have JSON along these lines; 

```
{
  "Network": {
      "id" : "sdfgh",
"network": {
      "nodes": [
        {
          "id": "ed228e09-d0bd-4f5a-a0c1-0eed3a8a783b",
          "name": "",
          "type": "Exposure",
          "payload": {
            "ExposureSegments": null
          }
        },
        {
          "id": "93d00412-4db0-4606-ab72-8314f75816a9",
          "name": "Xl",
          "type": "Section",
          "payload": {
            "InceptionDate": "2019-01-01",
            "ExpiryDate": "2019-12-31",
            "Layers": [
              {               
                "occType": "Ded",
                "share": 1.0
              }
            ]
          }
        }
      ]
    }
  }
}

```
I'm getting this error; 
```
AirstreamError.scala:74 Received unknown response: status 400, body:
{"payload":{"path":".Network.network.nodes.0.payload","expected":"discriminated-union","message":"Unable to find discriminator type
```
Below is the attempt at my encoding. 

The JSON itself is minimised from from an external service, that I hope hoping to call via a smithy client... 
```
structure IlsopCfnExportItem {
    @httpLabel
    @required
    id: CfnId,
    network: CfnDigraph
}

structure CfnDigraph {
    @required
    nodes: CfnDigraphNodes,

}


list CfnDigraphNodes {    
    member: CfnDigraphNode
}

structure CfnDigraphNode {
    @required
    id: CfnNodeId,
    @required
    name: String,
    @required
    type: String,
    @required
    payload:CfnDigraphADT
}

@discriminated("type")
union CfnDigraphADT{
    Section: SectionPayload,
    Exposure: ExposurePayload,
    LayeredCover: LcPayload    
}

structure CfnExposureSubNode{
    ExposureSegments: EmptyArrayForNow
}

structure CfnLayeredCoverSubNode{
    Layers: Layers
}

structure SectionPayload{
    payload : CfnSectionSubNode
}

structure ExposurePayload{
    payload : CfnExposureSubNode
}

structure LcPayload{
    payload : CfnLayeredCoverSubNode
}


structure CfnSectionSubNode{
    InceptionDate: IlsLocalDate,
    ExpiryDate: IlsLocalDate,
    Layers: Layers
}

```

Based on my reading of this docs, this is a faithful encoding... but have I missed something?

**Comment by Baccata** [comment, marked as answer]
@Quafadas you have the `@discriminated` trait at the wrong level, or the `type` field in the json at the wrong level, depending on how you see things. 

1. You should remove the explicit `type` field you've added in the `CfnDigraphNode` shape in your smithy model. The `type` field in the json level is taken care of implicitly by smithy4s via the `discriminated` trait, without polluting the models. Smithy4s actually protects against the manual addition of the discriminator field, this is because giving the user access to the field would mean putting them at risk of creating invalid instances. 
2. The location of `@discriminated("type")` trait, as you've placed it, indicates that the json objects under `payload` should contain the `type` field. In other words, a valid json payload would be : 

```json
{
  "Network": {
      "id" : "sdfgh",
"network": {
      "nodes": [
        {
          "id": "ed228e09-d0bd-4f5a-a0c1-0eed3a8a783b",
          "name": "",
          "payload": {
            "type": "Exposure",
            "ExposureSegments": null
          }
        },
        {
          "id": "93d00412-4db0-4606-ab72-8314f75816a9",
          "name": "Xl",
          "payload": {
            "type": "Section",
            "InceptionDate": "2019-01-01",
            "ExpiryDate": "2019-12-31",
            "Layers": [
              {               
                "occType": "Ded",
                "share": 1.0
              }
            ]
          }
        }
      ]
    }
  }
}
```

**Reply by Quafadas** [reply]
Thankyou again for being willing to take the time. 

I think I got really close to this. I recall at some point, getting an error about the existence of the "type" field, but failed to understand it's significance.

I have this working now ... which is wonderful, thankyou!

**Reply by kubukoz** [reply]
I keep forgetting that `type` is an implicit field. I actually didn't see anything wrong with the original code 😬

**Comment by Baccata** [comment]
I'm also sorry to inform you that, in the current version of smithy4s, you'll have to duplicate some information in various structures : 

```kotlin 
@discriminated("type") 
union CfnDigraphNode {
    Section: SectionNode, 
    Exposure: ExposureNode, 
    LayeredCover: LayeredCoverNode 
}  

structure SectionNode {
    @required
    id: CfnNodeId,
    @required
    name: String,
    @required
    payload:SectionPayload
}

// ... etc 
```

On the brightside, we're working hard to make Smithy4s understand [Smithy 2.0](https://aws.amazon.com/blogs/developer/introducing-smithy-idl-2-0/) which was released last Wednesday. Smithy 2.0 adds the concept of "mixins" which will allow for saving a lot of duplication. We're hoping to get a release out in the weeks to come.

**Reply by Quafadas** [reply]
I am happy, to live with this :-).
